tasks/bash.sh

```diff
@@ -1,106 +1,153 @@
+  VERSION_ID="${release#* }"
+}
+
+# Last resort
+_uname() {
+  [[ $ID ]] || ID="$(uname)"
+  [[ $full ]] || full="$(uname -r)"
+}
+
+munge_name() {
+  if [[ ${!osmap[@]} =~ "${1,,}" ]]; then
+    echo "${osmap[${1,,}]}"
+  else
+    echo "${1^}"
+  fi
+}
+
+_tmp="$(mktemp)"
+exec 2>>"$_tmp"
+
+# Use indirection to munge PT_ environment variables
+# e.g. "$PT_version" becomes "$version"
+for v in ${!PT_*}; do
+  declare "${v#*PT_}"="${!v}"
+done
+
+# Set up an error trap and let any functions inhert it
+trap 'fail "Error getting system information"' ERR
+set -E
+
+# Taken from https://github.com/puppetlabs/facter/blob/master/lib/inc/facter/facts/os.hpp
+# If not in this list, we just uppercase the first character
+declare -A osmap=([redhat]=RedHat [rhel]=RedHat [centos]=CentOS [cloud]=CloudLinux [virtuozzo]=VirtuozzoLinux [psbm]=PSBM [ol]=OracleLinux [ovl]=OSV [oel]=OEL [xenserver]=XenServer [linuxmint]=LinuxMint [sles]=SLES [suse]=SuSE [opensuse]=OpenSuSE [sunos]=SunOS [omni]=OmniOS [openindiana]=OpenIndiana [manjaro]=ManjaroLinux [smart]=SmartOS [openwrt]=OpenWrt [meego]=MeeGo [coreos]=CoreOS [zen]=XCP [kfreebsd]='GNU/kFreeBSD' [arista]=AristaEOS [huawei]=HuaweiOS [photon]=PhotonOS)
+
+if [[ -e /etc/os-release ]]; then
+  _systemd /etc/os-release
+elif [[ -e /usr/lib/os-release ]]; then
+  _systemd /usr/lib/os-release
 fi
 
-if [ -z "${name}" ]; then
-    name=$(uname)
-    release=$(uname -r)
+# If either systemd is not installed or we didn't get a minor version from os-release
+if ! [[ $VERSION_ID ]] || (( ${VERSION_ID%%.*} == ${VERSION_ID#*.} )); then
+  if [[ -e /etc/fedora-release ]]; then
+    _rhel /etc/fedora-release
+  elif [[ -e /etc/centos-release ]]; then
+    _rhel /etc/centos-release
+  elif [[ -e /etc/redhat-release ]]; then
+    _rhel /etc/redhat-release
+  elif type lsb_release &>/dev/null; then
+    _lsb_release
+  else
+    _uname
+  fi
 fi
 
-# puppet_agent install task can ask for each of these values
-if test "x$1" = "xplatform"; then
-    echo $name
-    exit 0
-elif test "x$1" = "xrelease"; then
-    echo $release
-    exit 0
+full="${VERSION_ID}"
+major="${VERSION_ID%%.*}"
+# Minor is considered the second part of the version string
+IFS='.' read -ra minor <<<"$full"
+minor="${minor[1]}"
+
+ID="$(munge_name "$ID")"
+family="$(munge_name "$family")"
+
+# We should change puppet_agent to not work this way
+if [[ $@ =~ 'platform' ]]; then
+  success "$ID"
+elif [[ $@ =~ 'release' ]]; then
+  success "$full"
 fi
 
-case $name in
-    RedHat|Fedora|CentOS|Scientific|SLC|Ascendos|CloudLinux)
-        family=RedHat;;
-    HuaweiOS|LinuxMint|Ubuntu|Debian)
-        family=Debian;;
-    *)
-        family=$name;;
-esac
-
-# Print it all out
-if [ -z "$name" ]; then
-    cat <<JSON
-{
-  "_error": {
-    "kind": "facts/noname",
-    "msg": "Could not determine OS name"
-  }
-}
-JSON
-else
-    cat <<JSON
+success "$(cat <<EOF
 {
   "os": {
-    "name": "${name}",
-JSON
-    [ -n "$release" ] && cat <<JSON
+    "name": "$ID",
```

Comment:
For some of the older Redhat platforms ( centos, redhat, oracle) the ID is lost. See https://jenkins-master-prod-1.delivery.puppetlabs.net/job/forge-module_puppetlabs-facts_intn-sys_full-pe-current-adhoc/ for a list. 
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt/modules$ bolt task run facts::bash -t ua1gehq2djh63eu.delivery.puppetlabs.net
Started on ua1gehq2djh63eu.delivery.puppetlabs.net...
Finished on ua1gehq2djh63eu.delivery.puppetlabs.net:
  {
    "os": {
      "name": "",
      "codename": "Santiago",
      "release": {
        "full": "6.8",
        "major": "6",
        "minor": "8"
      },
      "family": "RedHat"
    }
  }
Successful on 1 node: ua1gehq2djh63eu.delivery.puppetlabs.net
Ran on 1 node in 0.47 seconds
```

Reply:
Thanks, I'll look into this.

Reply:
Ah, it's because the associative array declaration doesn't work on those version of Bash.  I'll think of something...